// @colyseus/core/build/packages/core/src/rooms/RelayRoom.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../../../node_modules/tslib/tslib.es6.js');
var schema = require('@colyseus/schema');
var Room = require('../Room.js');

/**
 * Create another context to avoid these types from being in the user's global `Context`
 */
const context = new schema.Context();
class Player extends schema.Schema {
}
schema.defineTypes(Player, {
    connected: 'boolean',
    name: 'string',
    sessionId: 'string',
}, context);
class State extends schema.Schema {
    constructor() {
        super(...arguments);
        this.players = new schema.MapSchema();
    }
}
schema.defineTypes(State, {
    players: { map: Player },
}, context);
/**
 * client.joinOrCreate("relayroom", {
 *   maxClients: 10,
 *   allowReconnectionTime: 20
 * });
 */
class RelayRoom extends Room.Room {
    constructor() {
        super(...arguments);
        this.allowReconnectionTime = 0;
    }
    onCreate(options) {
        this.setState(new State());
        if (options.maxClients) {
            this.maxClients = options.maxClients;
        }
        if (options.allowReconnectionTime) {
            this.allowReconnectionTime = Math.min(options.allowReconnectionTime, 40);
        }
        if (options.metadata) {
            this.setMetadata(options.metadata);
        }
        this.onMessage('*', (client, type, message) => {
            this.broadcast(type, [client.sessionId, message], { except: client });
        });
    }
    onJoin(client, options = {}) {
        const player = new Player();
        player.connected = true;
        player.sessionId = client.sessionId;
        if (options.name) {
            player.name = options.name;
        }
        this.state.players.set(client.sessionId, player);
    }
    onLeave(client, consented) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            if (this.allowReconnectionTime > 0) {
                const player = this.state.players.get(client.sessionId);
                player.connected = false;
                try {
                    if (consented) {
                        throw new Error('consented leave');
                    }
                    yield this.allowReconnection(client, this.allowReconnectionTime);
                    player.connected = true;
                }
                catch (e) {
                    this.state.players.delete(client.sessionId);
                }
            }
        });
    }
}

exports.RelayRoom = RelayRoom;
