// @colyseus/core/build/packages/core/src/MatchMaker.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../../node_modules/tslib/tslib.es6.js');
var Protocol = require('./Protocol.js');
var IPC = require('./IPC.js');
var Utils = require('./Utils.js');
var RegisteredHandler = require('./matchmaker/RegisteredHandler.js');
var Room = require('./Room.js');
var LocalPresence = require('./presence/LocalPresence.js');
var Debug = require('./Debug.js');
var SeatReservationError = require('./errors/SeatReservationError.js');
var ServerError = require('./errors/ServerError.js');
var index = require('./matchmaker/driver/index.js');
var controller = require('./matchmaker/controller.js');

const handlers = {};
const rooms = {};
exports.processId = void 0;
exports.presence = void 0;
exports.driver = void 0;
let isGracefullyShuttingDown;
function setup(_presence, _driver, _processId) {
    exports.presence = _presence || new LocalPresence.LocalPresence();
    exports.driver = _driver || new index.LocalDriver();
    exports.processId = _processId;
    isGracefullyShuttingDown = false;
    /**
     * Subscribe to remote `handleCreateRoom` calls.
     */
    IPC.subscribeIPC(exports.presence, exports.processId, getProcessChannel(), (_, args) => {
        return handleCreateRoom.apply(undefined, args);
    });
    exports.presence.hset(getRoomCountKey(), exports.processId, '0');
}
/**
 * Join or create into a room and return seat reservation
 */
function joinOrCreate(roomName, clientOptions = {}) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        return yield Utils.retry(() => tslib_es6.__awaiter(this, void 0, void 0, function* () {
            let room = yield findOneRoomAvailable(roomName, clientOptions);
            if (!room) {
                room = yield createRoom(roomName, clientOptions);
            }
            return yield reserveSeatFor(room, clientOptions);
        }), 5, [SeatReservationError.SeatReservationError]);
    });
}
/**
 * Create a room and return seat reservation
 */
function create(roomName, clientOptions = {}) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        const room = yield createRoom(roomName, clientOptions);
        return reserveSeatFor(room, clientOptions);
    });
}
/**
 * Join a room and return seat reservation
 */
function join(roomName, clientOptions = {}) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        return yield Utils.retry(() => tslib_es6.__awaiter(this, void 0, void 0, function* () {
            const room = yield findOneRoomAvailable(roomName, clientOptions);
            if (!room) {
                throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_INVALID_CRITERIA, `no rooms found with provided criteria`);
            }
            return reserveSeatFor(room, clientOptions);
        }));
    });
}
/**
 * Join a room by id and return seat reservation
 */
function joinById(roomId, clientOptions = {}) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        const room = yield exports.driver.findOne({ roomId });
        if (room) {
            const rejoinSessionId = clientOptions.sessionId;
            if (rejoinSessionId) {
                // handle re-connection!
                const hasReservedSeat = yield remoteRoomCall(room.roomId, 'hasReservedSeat', [rejoinSessionId]);
                if (hasReservedSeat) {
                    return { room, sessionId: rejoinSessionId };
                }
                else {
                    throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_EXPIRED, `session expired: ${rejoinSessionId}`);
                }
            }
            else if (!room.locked) {
                return reserveSeatFor(room, clientOptions);
            }
            else {
                throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_INVALID_ROOM_ID, `room "${roomId}" is locked`);
            }
        }
        else {
            throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_INVALID_ROOM_ID, `room "${roomId}" not found`);
        }
    });
}
/**
 * Perform a query for all cached rooms
 */
function query(conditions = {}) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        return yield exports.driver.find(conditions);
    });
}
/**
 * Find for a public and unlocked room available
 */
function findOneRoomAvailable(roomName, clientOptions) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        return yield awaitRoomAvailable(roomName, () => tslib_es6.__awaiter(this, void 0, void 0, function* () {
            const handler = handlers[roomName];
            if (!handler) {
                throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_NO_HANDLER, `provided room name "${roomName}" not defined`);
            }
            const roomQuery = exports.driver.findOne(Object.assign({ locked: false, name: roomName, private: false }, handler.getFilterOptions(clientOptions)));
            if (handler.sortOptions) {
                roomQuery.sort(handler.sortOptions);
            }
            return yield roomQuery;
        }));
    });
}
/**
 * Call a method or return a property on a remote room.
 */
function remoteRoomCall(roomId, method, args, rejectionTimeout = Utils.REMOTE_ROOM_SHORT_TIMEOUT) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        const room = rooms[roomId];
        if (!room) {
            try {
                return yield IPC.requestFromIPC(exports.presence, getRoomChannel(roomId), method, args);
            }
            catch (e) {
                const request = `${method}${args && ' with args ' + JSON.stringify(args) || ''}`;
                throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_UNHANDLED, `remote room (${roomId}) timed out, requesting "${request}". (${rejectionTimeout}ms exceeded)`);
            }
        }
        else {
            return (!args && typeof (room[method]) !== 'function')
                ? room[method]
                : (yield room[method].apply(room, args));
        }
    });
}
function defineRoomType(name, klass, defaultOptions) {
    const registeredHandler = new RegisteredHandler.RegisteredHandler(klass, defaultOptions);
    handlers[name] = registeredHandler;
    cleanupStaleRooms(name);
    return registeredHandler;
}
function removeRoomType(name) {
    delete handlers[name];
    cleanupStaleRooms(name);
}
function hasHandler(name) {
    return handlers[name] !== undefined;
}
/**
 * Create a room
 */
function createRoom(roomName, clientOptions) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        const roomsSpawnedByProcessId = yield exports.presence.hgetall(getRoomCountKey());
        const processIdWithFewerRooms = (Object.keys(roomsSpawnedByProcessId).sort((p1, p2) => {
            return (Number(roomsSpawnedByProcessId[p1]) > Number(roomsSpawnedByProcessId[p2]))
                ? 1
                : -1;
        })[0]) || exports.processId;
        if (processIdWithFewerRooms === exports.processId) {
            // create the room on this process!
            return yield handleCreateRoom(roomName, clientOptions);
        }
        else {
            // ask other process to create the room!
            let room;
            try {
                room = yield IPC.requestFromIPC(exports.presence, getProcessChannel(processIdWithFewerRooms), undefined, [roomName, clientOptions], Utils.REMOTE_ROOM_SHORT_TIMEOUT);
            }
            catch (e) {
                // if other process failed to respond, create the room on this process
                Debug.debugAndPrintError(e);
                room = yield handleCreateRoom(roomName, clientOptions);
            }
            return room;
        }
    });
}
function handleCreateRoom(roomName, clientOptions) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        const registeredHandler = handlers[roomName];
        if (!registeredHandler) {
            throw new ServerError.ServerError(Protocol.ErrorCode.MATCHMAKE_NO_HANDLER, `provided room name "${roomName}" not defined`);
        }
        const room = new registeredHandler.klass();
        // set room public attributes
        room.roomId = Utils.generateId();
        room.roomName = roomName;
        room.presence = exports.presence;
        // create a RoomCache reference.
        room.listing = exports.driver.createInstance(Object.assign({ name: roomName, processId: exports.processId }, registeredHandler.getFilterOptions(clientOptions)));
        if (room.onCreate) {
            try {
                yield room.onCreate(Utils.merge({}, clientOptions, registeredHandler.options));
                // increment amount of rooms this process is handling
                exports.presence.hincrby(getRoomCountKey(), exports.processId, 1);
            }
            catch (e) {
                Debug.debugAndPrintError(e);
                throw new ServerError.ServerError(e.code || Protocol.ErrorCode.MATCHMAKE_UNHANDLED, e.message);
            }
        }
        room.internalState = Room.RoomInternalState.CREATED;
        room.listing.roomId = room.roomId;
        room.listing.maxClients = room.maxClients;
        // imediatelly ask client to join the room
        Debug.debugMatchMaking('spawning \'%s\', roomId: %s, processId: %s', roomName, room.roomId, exports.processId);
        room._events.on('lock', lockRoom.bind(this, room));
        room._events.on('unlock', unlockRoom.bind(this, room));
        room._events.on('join', onClientJoinRoom.bind(this, room));
        room._events.on('leave', onClientLeaveRoom.bind(this, room));
        room._events.once('dispose', disposeRoom.bind(this, roomName, room));
        room._events.once('disconnect', () => room._events.removeAllListeners());
        // room always start unlocked
        yield createRoomReferences(room, true);
        yield room.listing.save();
        registeredHandler.emit('create', room);
        return room.listing;
    });
}
function getRoomById(roomId) {
    return rooms[roomId];
}
function gracefullyShutdown() {
    if (isGracefullyShuttingDown) {
        return Promise.reject('already_shutting_down');
    }
    isGracefullyShuttingDown = true;
    Debug.debugMatchMaking(`${exports.processId} is shutting down!`);
    // remove processId from room count key
    exports.presence.hdel(getRoomCountKey(), exports.processId);
    // unsubscribe from process id channel
    exports.presence.unsubscribe(getProcessChannel());
    const promises = [];
    for (const roomId in rooms) {
        if (!rooms.hasOwnProperty(roomId)) {
            continue;
        }
        promises.push(rooms[roomId].disconnect());
    }
    return Promise.all(promises);
}
/**
 * Reserve a seat for a client in a room
 */
function reserveSeatFor(room, options) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        const sessionId = Utils.generateId();
        Debug.debugMatchMaking('reserving seat. sessionId: \'%s\', roomId: \'%s\', processId: \'%s\'', sessionId, room.roomId, exports.processId);
        let successfulSeatReservation;
        try {
            successfulSeatReservation = yield remoteRoomCall(room.roomId, '_reserveSeat', [sessionId, options]);
        }
        catch (e) {
            Debug.debugMatchMaking(e);
            successfulSeatReservation = false;
        }
        if (!successfulSeatReservation) {
            throw new SeatReservationError.SeatReservationError(`${room.roomId} is already full.`);
        }
        return { room, sessionId };
    });
}
function cleanupStaleRooms(roomName) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        //
        // clean-up possibly stale room ids
        // (ungraceful shutdowns using Redis can result on stale room ids still on memory.)
        //
        const cachedRooms = yield exports.driver.find({ name: roomName }, { _id: 1 });
        // remove connecting counts
        yield exports.presence.del(getHandlerConcurrencyKey(roomName));
        yield Promise.all(cachedRooms.map((room) => tslib_es6.__awaiter(this, void 0, void 0, function* () {
            try {
                // use hardcoded short timeout for cleaning up stale rooms.
                yield remoteRoomCall(room.roomId, 'roomId');
            }
            catch (e) {
                Debug.debugMatchMaking(`cleaning up stale room '${roomName}', roomId: ${room.roomId}`);
                room.remove();
            }
        })));
    });
}
function createRoomReferences(room, init = false) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        rooms[room.roomId] = room;
        if (init) {
            yield IPC.subscribeIPC(exports.presence, exports.processId, getRoomChannel(room.roomId), (method, args) => {
                return (!args && typeof (room[method]) !== 'function')
                    ? room[method]
                    : room[method].apply(room, args);
            });
        }
        return true;
    });
}
function awaitRoomAvailable(roomToJoin, callback) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => tslib_es6.__awaiter(this, void 0, void 0, function* () {
            const concurrencyKey = getHandlerConcurrencyKey(roomToJoin);
            const concurrency = (yield exports.presence.incr(concurrencyKey)) - 1;
            // avoid having too long timeout if 10+ clients ask to join at the same time
            const concurrencyTimeout = Math.min(concurrency * 100, Utils.REMOTE_ROOM_SHORT_TIMEOUT);
            if (concurrency > 0) {
                Debug.debugMatchMaking('receiving %d concurrent requests for joining \'%s\' (waiting %d ms)', concurrency, roomToJoin, concurrencyTimeout);
            }
            setTimeout(() => tslib_es6.__awaiter(this, void 0, void 0, function* () {
                try {
                    const result = yield callback();
                    resolve(result);
                }
                catch (e) {
                    reject(e);
                }
                finally {
                    yield exports.presence.decr(concurrencyKey);
                }
            }), concurrencyTimeout);
        }));
    });
}
function onClientJoinRoom(room, client) {
    handlers[room.roomName].emit('join', room, client);
}
function onClientLeaveRoom(room, client, willDispose) {
    handlers[room.roomName].emit('leave', room, client, willDispose);
}
function lockRoom(room) {
    // emit public event on registered handler
    handlers[room.roomName].emit('lock', room);
}
function unlockRoom(room) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        if (yield createRoomReferences(room)) {
            // emit public event on registered handler
            handlers[room.roomName].emit('unlock', room);
        }
    });
}
function disposeRoom(roomName, room) {
    return tslib_es6.__awaiter(this, void 0, void 0, function* () {
        Debug.debugMatchMaking('disposing \'%s\' (%s) on processId \'%s\'', roomName, room.roomId, exports.processId);
        // decrease amount of rooms this process is handling
        if (!isGracefullyShuttingDown) {
            exports.presence.hincrby(getRoomCountKey(), exports.processId, -1);
        }
        // remove from room listing (already removed if `disconnect()` has been called)
        if (room.internalState !== Room.RoomInternalState.DISCONNECTING) {
            yield room.listing.remove();
        }
        // emit disposal on registered session handler
        handlers[roomName].emit('dispose', room);
        // remove concurrency key
        exports.presence.del(getHandlerConcurrencyKey(roomName));
        // unsubscribe from remote connections
        exports.presence.unsubscribe(getRoomChannel(room.roomId));
        // remove actual room reference
        delete rooms[room.roomId];
    });
}
//
// Presence keys
//
function getRoomChannel(roomId) {
    return `$${roomId}`;
}
function getHandlerConcurrencyKey(name) {
    return `c:${name}`;
}
function getProcessChannel(id = exports.processId) {
    return `p:${id}`;
}
function getRoomCountKey() {
    return 'roomcount';
}

exports.controller = controller;
exports.create = create;
exports.createRoom = createRoom;
exports.defineRoomType = defineRoomType;
exports.findOneRoomAvailable = findOneRoomAvailable;
exports.getRoomById = getRoomById;
exports.gracefullyShutdown = gracefullyShutdown;
exports.hasHandler = hasHandler;
exports.join = join;
exports.joinById = joinById;
exports.joinOrCreate = joinOrCreate;
exports.query = query;
exports.remoteRoomCall = remoteRoomCall;
exports.removeRoomType = removeRoomType;
exports.reserveSeatFor = reserveSeatFor;
exports.setup = setup;
